pkg/controller/ksmtuned/controller.go

```diff
@@ -0,0 +1,159 @@
+	ctlksmtuned "github.com/harvester/node-manager/pkg/generated/controllers/harvesterhci.io/v1beta1"
+	"github.com/harvester/node-manager/pkg/ksmtuned"
+	"github.com/harvester/node-manager/pkg/metrics"
+)
+
+const (
+	HandlerName     = "harvester-ksmtuned-handler"
+	NodeHandlerName = "harvester-ksmtuned-node-handler"
+)
+
+var (
+	modes = map[ksmtunedv1.KsmtunedMode]*ksmtunedv1.KsmtunedParameters{
+		ksmtunedv1.StandardMode: {
+			SleepMsec: 20,
+			Boost:     0,
+			Decay:     0,
+			MinPages:  100,
+			MaxPages:  100,
+		},
+		ksmtunedv1.HighMode: {
+			SleepMsec: 20,
+			Boost:     200,
+			Decay:     50,
+			MinPages:  100,
+			MaxPages:  10000,
+		},
+	}
+)
+
+type Controller struct {
+	ctx      context.Context
+	NodeName string
+
+	KsmtunedCache ctlksmtuned.KsmtunedCache
+	Ksmtuneds     ctlksmtuned.KsmtunedController
+
+	NodeCache ctlnode.NodeCache
+	Nodes     ctlnode.NodeController
+
+	Ksmtuned *ksmtuned.Ksmtuned
+
+	CurrentKsmtuendSpec *ksmtunedv1.KsmtunedSpec
+}
+
+func Register(ctx context.Context, nodeName string, kts ctlksmtuned.KsmtunedController, nodes ctlnode.NodeController) (*Controller, error) {
+	k, err := ksmtuned.NewKsmtuned(ctx, nodeName)
+	if err != nil {
+		return nil, err
+	}
+
+	k.SetKsmdUtilization(metrics.KsmdUtilizationGV)
+
+	c := &Controller{
+		ctx:           ctx,
+		NodeName:      nodeName,
+		KsmtunedCache: kts.Cache(),
+		Ksmtuneds:     kts,
+		NodeCache:     nodes.Cache(),
+		Nodes:         nodes,
+		Ksmtuned:      k,
+	}
+
+	c.Ksmtuneds.OnChange(ctx, HandlerName, c.OnChange)
+	c.Ksmtuneds.OnRemove(ctx, HandlerName, c.OnRemove)
+
+	c.Nodes.OnChange(ctx, NodeHandlerName, c.NodeOnChange)
+	c.Nodes.OnRemove(ctx, NodeHandlerName, c.NodeOnRemove)
+
+	go c.watchStatus(nodeName)
+
+	return c, nil
+}
+
+func (c *Controller) OnChange(key string, kt *ksmtunedv1.Ksmtuned) (*ksmtunedv1.Ksmtuned, error) {
+	if kt == nil || kt.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	if key != c.NodeName {
+		return kt, nil
+	}
+
+	if c.CurrentKsmtuendSpec == nil || !reflect.DeepEqual(c.CurrentKsmtuendSpec, &kt.Spec) {
+		c.CurrentKsmtuendSpec = &kt.Spec
+	} else {
+		// not any changes
+		return kt, nil
+	}
+
+	var (
+		parameters *ksmtunedv1.KsmtunedParameters
+		ok         bool
+	)
+
+	if !kt.Spec.Enable {
+		parameters = nil
+	} else if parameters, ok = modes[kt.Spec.Mode]; !ok { // customized mode
+		parameters = kt.Spec.CustomizedParameters
+	}
+
+	if err := c.Ksmtuned.Apply(kt.Spec.ThresCoef, parameters); err != nil {
+		return nil, err
+	}
+
+	return kt, nil
+}
+
+func (c *Controller) OnRemove(_ string, kt *ksmtunedv1.Ksmtuned) (*ksmtunedv1.Ksmtuned, error) {
+	if kt.Name != c.NodeName {
+		return kt, nil
+	}
+	return kt, c.Ksmtuned.Stop()
+}
+
+func (c *Controller) watchStatus(name string) {
+	var lastStatusUpdated ksmtunedv1.KsmtunedStatus
+
+	ch := c.Ksmtuned.Status()
+	for s := range ch {
+		if reflect.DeepEqual(lastStatusUpdated, s) {
+			continue
+		}
```

Comment:
this check feels to be redundant, perfer to get from cache and compare its status below directly.